File: pyclustering/cluster/hierarchical.py
"""!

@brief Cluster analysis algorithm: hierarchical algorithm.
@details Implementation based on article:
         - K.Anil, J.C.Dubes, R.C.Dubes. Algorithms for Clustering Data. 1988.

@authors Andrei Novikov (spb.andr@yandex.ru)
@version 1.0
@date 2014-2015
@copyright GNU Public License

@cond GNU_PUBLIC_LICENSE
    PyClustering is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.
    
    PyClustering is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    
    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
@endcond

"""

from pyclustering.support import euclidean_distance_sqrt;

import pyclustering.core.wrapper as wrapper;


class hierarchical:
    """!
    @brief Class represents hierarchical algorithm for cluster analysis.
    
    Example:
    @code
        # sample for cluster analysis (represented by list)
        sample = read_sample(path_to_sample);
        
        # create object that uses python code only
        hierarchical_instance = hierarchical(sample, 2, False)
        
        # cluster analysis
        hierarchical_instance.process();
        
        # obtain results of clustering
        clusters = hierarchical_instance.get_clusters();  
    @endcode
    
    """
        
    __pointer_data = None;
    __number_clusters = 0;
    
    __ccore = False;
    
    __clusters = None;
    __centers = None;
    
    def __init__(self, data, number_clusters, ccore):
        """!
        @brief Constructor of clustering algorithm hierarchical.
        
        @param[in] data (list): Input data that is presented as a list of points (objects), each point should be represented by a list or tuple.
        @param[in] number_clusters (uint): Number of clusters that should be allocated.
        @param[in] ccore (bool): if True than DLL CCORE (C++ solution) will be used for solving the problem.
        
        """  
        self.__pointer_data = data;
        self.__number_clusters = number_clusters;
        self.__ccore = None;
        
        self.__clusters = [];
        self.__centers = [];
        
    
    def process(self):
        """!
        @brief Performs cluster analysis in line with rules of hierarchical algorithm.
        
        @see get_clusters()
        
        """
        if (self.__ccore is True):
            self.__clusters = wrapper.hierarchical(self.__pointer_data, self.__number_clusters); 
        else:        
            self.__centers = self.__pointer_data.copy();
            self.__clusters = [[index] for index in range(0, len(self.__pointer_data))];
            
            current_number_clusters = len(self.__clusters);
            
            while (current_number_clusters > self.__number_clusters):
                indexes = self.__find_nearest_clusters();
                 
                self.__clusters[indexes[0]] += self.__clusters[indexes[1]];
                self.__centers[indexes[0]] = self.__calculate_center(self.__clusters[indexes[0]]);
                 
                self.__clusters.pop(indexes[1]);   # remove merged cluster.
                self.__centers.pop(indexes[1]);    # remove merged center.
                
                current_number_clusters = len(self.__clusters);
        
        
    def get_clusters(self):
        """!
        @brief Performs cluster analysis in line with rules of heirarchical algorithm.
        
        @remark Results of clustering can be obtained using corresponding gets methods.
        
        @return (list) List of allocated clusters, each cluster contains indexes of objects in list of data.
        
        @see process()
        
        """
        
        return self.__clusters;


    def __find_nearest_clusters(self):
        """!
        @brief Find two indexes of two clusters whose distance is the smallest.
        
        @return (list) List with two indexes of two clusters whose distance is the smallest.
        
        """     
        
        min_dist = 0;
        indexes = None;
        
        for index1 in range(0, len(self.__centers)):
            for index2 in range(index1 + 1, len(self.__centers)):
                distance = euclidean_distance_sqrt(self.__centers[index1], self.__centers[index2]);
                if ( (distance < min_dist) or (indexes == None) ):
                    min_dist = distance;
                    indexes = [index1, index2];
        
        return indexes; 
    
           
    def __calculate_center(self, cluster):
        """!
        @brief Calculates new center.
        
        @return (list) New value of the center of the specified cluster.
        
        """
         
        dimension = len(self.__pointer_data[cluster[0]]);
        center = [0] * dimension;
        for index_point in cluster:
            for index_dimension in range(0, dimension):
                center[index_dimension] += self.__pointer_data[index_point][index_dimension];
         
        for index_dimension in range(0, dimension):
            center[index_dimension] /= len(cluster);
             
        return center;

File: ccore/ccore/hierarchical.cpp
/**************************************************************************************************************

Cluster analysis algorithm: Classical Hierarchical Algorithm

Based on article description:
 - K.Anil, J.C.Dubes, R.C.Dubes. Algorithms for Clustering Data. 1988.

Copyright (C) 2015    Andrei Novikov (pyclustering@yandex.ru)

pyclustering is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

pyclustering is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

**************************************************************************************************************/

#include "hierarchical.h"

#include "support.h"

#if 1
#define FAST_SOLUTION
#endif

hierarchical_cluster::hierarchical_cluster(const std::vector<std::vector<double> > * const data, const unsigned int index, const std::vector<double> * const point) {
	dataset = (std::vector<std::vector<double> > *) data;
	center = new std::vector<double>(*point);
	indexes = new std::vector<unsigned int>(1, index);
}

hierarchical_cluster::~hierarchical_cluster(void) {
	dataset = NULL;

	if (center != NULL) {
		delete center; 
		center = NULL;
	}

	if (indexes != NULL) {
		delete indexes; 
		indexes = NULL;
	}
}

void hierarchical_cluster::append(const hierarchical_cluster * const cluster2) {
	indexes->insert(indexes->end(), cluster2->get_indexes()->begin(), cluster2->get_indexes()->end());

	/* update center */
	for (unsigned int dimension = 0; dimension < center->size(); dimension++) {
		(*center)[dimension] = 0.0;
		for (std::vector<unsigned int>::const_iterator iter = indexes->begin(); iter != indexes->end(); iter++) {
			(*center)[dimension] += (*dataset)[*iter][dimension];
		}

		(*center)[dimension] /= indexes->size();
	}
}

hierarchical::hierarchical(const std::vector<std::vector<double> > * const dataset, unsigned int cluster_number) {
	data = (std::vector<std::vector<double> > *) dataset;
	clusters = new std::list<hierarchical_cluster *>();

	standard_clusters = NULL;

	number_clusters = cluster_number;

	for (unsigned int index = 0; index < dataset->size(); index++) {
		hierarchical_cluster * data_cluster = new hierarchical_cluster(dataset, index, &(dataset->data()[index]));
		clusters->push_back(data_cluster);
	}
}

hierarchical::~hierarchical(void) {
	data = NULL;

	if (clusters != NULL) {
		for (std::list<hierarchical_cluster *>::const_iterator iter = clusters->begin(); iter != clusters->end(); iter++) {
			delete (*iter);
		}

		delete clusters;
		clusters = NULL;
	}

	if (standard_clusters != NULL) {
		delete standard_clusters;
		standard_clusters = NULL;
	}
}

void hierarchical::process(void) {
	while(clusters->size() > number_clusters) {
		merge_nearest_clusters();
	}

	if (standard_clusters != NULL) {
		delete standard_clusters;
		standard_clusters = NULL;
	}

	if (clusters->size() > 0) {
		standard_clusters = new std::vector<std::vector<unsigned int> * >();
		for (std::list<hierarchical_cluster *>::const_iterator iter = clusters->begin(); iter != clusters->end(); iter++) {
			standard_clusters->push_back( (std::vector<unsigned int> *) (*iter)->get_indexes() );
		}
	}
}

void hierarchical::merge_nearest_clusters(void) {
	double minimum_distance = INFINITY;
	std::list<hierarchical_cluster *>::iterator candidate1, candidate2;

	for(std::list<hierarchical_cluster *>::iterator iter1 = clusters->begin(); iter1 != clusters->end(); iter1++) {
		std::list<hierarchical_cluster *>::iterator iter2 = iter1;
		for(++iter2; iter2 != clusters->end(); iter2++) {
#ifndef FAST_SOLUTION
			double distance = euclidean_distance_sqrt((*iter1)->get_center(), (*iter2)->get_center());
#else
			double distance = euclidean_distance((*iter1)->get_center(), (*iter2)->get_center());
#endif
			if (distance < minimum_distance) {
				candidate1 = iter1;
				candidate2 = iter2;
				minimum_distance = distance;
			}
		}
	}

	((hierarchical_cluster *) (*candidate1))->append(*candidate2);
	clusters->erase(candidate2);
}
